src/send-and-sync.md

```diff
@@ -74,7 +74,114 @@ of their pervasive use of raw pointers to manage allocations and complex ownersh
 Similarly, most iterators into these collections are Send and Sync because they
 largely behave like an `&` or `&mut` into the collection.
 
+[`Box`][box-doc] is implemented as it's own special intrinsic type by the
+compiler for [various reasons][box-is-special], but we can implement something
+with similar-ish behaviour ourselves to see an example of when it is sound to
+implement Send and Sync. Let's call it a `Carton`.
+
+We start by writing code to take a value allocated on the stack and transfer it
+to the heap.
+
+```rust,ignore
+use std::mem::size_of;
+use std::ptr::NonNull;
+
+struct Carton<T>(NonNull<T>);
+
+impl<T> Carton<T> {
+    pub fn new(mut value: T) -> Self {
+        // Allocate enough memory on the heap to store one T
+        let ptr = unsafe { libc::calloc(1, size_of::<T>()) as *mut T };
+
+        // NonNull is just a wrapper that enforces that the pointer isn't null.
+        // Malloc returns null if it can't allocate.
+        let mut ptr = NonNull::new(ptr).expect("We assume malloc doesn't fail");
+
+        // Move value from the stack to the location we allocated on the heap
+        unsafe {
+            // Safety: The pointer returned by calloc is alligned, initialized,
```

Comment:
I don't believe `calloc` always yields well-aligned pointers. AFAIK, `calloc` can yield addresses that are not aligned to `4 * size_of::<usize>`, for instance. That is, on a 32-bit machine, you may get pointers that are not `0x10`-aligned.

I thus suggest you put an assertion in there that `align_of::<T>` is small enough

Reply:
Thanks! I fixed this by using `posix_memalign` in c3ba178